### PR TITLE
[tensorflow] tf implementation of section 12.1, compilers and interpreters

### PR DESCRIPTION
**Pull request by levimcclenny**
Added tf implementation for section 12.1. I took a small liberty in the section on JIT compilation and added an example that compiles with XLA as well as standard graph-mode JIT. If we want it to be closer to the mxnet and pytorch versions then I can remove this. Additionally the final example of this section takes advantage of multi-dispatch functionality that I believe is exclusive to mxnet, so I did not do a tensorflow example for that. There was not a pytorch example either for this specific one.

Thanks,  

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by mli** [comment]
Job d2l-en/PR-1646/3 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1646/

**Comment by levimcclenny** [comment]
update: Had to remove the XLA example code, CI is in 2.3.1 and I believe `jit_compile` was added to tf.function in 2.4.X. I left the description of XLA in the write-up but removed the code examples so CI would pass.

**Comment by mli** [comment]
Job d2l-en/PR-1646/4 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1646/

File: chapter_computational-performance/hybridize.md [hunk 1]
```diff
@@ -79,6 +79,10 @@ In practice this means that we build models using either the `HybridBlock` or th
 As mentioned above, PyTorch is based on imperative programming and uses dynamic computation graphs. In an effort to leverage the portability and efficiency of symbolic programming, developers considered whether it would be possible to combine the benefits of both programming models. This led to a torchscript that lets users develop and debug using pure imperative programming, while having the ability to convert most programs into symbolic programs to be run when product-level computing performance and deployment are required.
 :end_tab:
 
+:begin_tab:`tensorflow`
+The imperative programming paradigm is now the default in Tensorflow 2.X, a welcome change for those new to the language. However, the same symbolic programming techniques and subsequent computational graphs still exist in tensorflow, and can be accessed by the easy-to-use `tf.function` decorator. This brought the imperative programming paradigm to tensorflow, allowed users to define more intuitive functions, then wrap them and compile them into computational graphs automatically using a feature the Tensorflow team refers to as [autograph](https://www.tensorflow.org/api_docs/python/tf/autograph).
```
Comment (terrytangyuan) [inline chapter_computational-performance/hybridize.md:83]:
```suggestion
The imperative programming paradigm is now the default in Tensorflow 2, a welcoming change for those new to the language. However, the same symbolic programming techniques and subsequent computational graphs still exist in TensorFlow, and can be accessed by the easy-to-use `tf.function` decorator. This brought the imperative programming paradigm to TensorFlow, allowed users to define more intuitive functions, then wrap them and compile them into computational graphs automatically using a feature the TensorFlow team refers to as [autograph](https://www.tensorflow.org/api_docs/python/tf/autograph).
```

File: chapter_computational-performance/hybridize.md [hunk 2]
```diff
@@ -150,6 +184,11 @@ This seems almost too good to be true: simply designate a block to be `HybridSeq
 By converting the model using `torch.jit.script` This seems almost too good to be true: write the same code as before and simply convert the model using `torch.jit.script`. Once this happens the network is optimized (we will benchmark the performance below).
 :end_tab:
 
+:begin_tab:`tensorflow`
+Converting the model using `tf.function` gives us incredible power in tensorflow: write the same code as before and simply convert the model using `tf.function`. Once this happens the network is built as a computational graph in tensorflow's MLIR intermediate representation and is heavily optimized at the compiler level for rapid execution (we will benchmark the performance below).
```
Comment (terrytangyuan) [inline chapter_computational-performance/hybridize.md:188]:
```suggestion
Converting the model using `tf.function` gives us incredible power in TensorFlow: write the same code as before and simply convert the model using `tf.function`. Once this happens the network is built as a computational graph in TensorFlow's MLIR intermediate representation and is heavily optimized at the compiler level for rapid execution (we will benchmark the performance below).
```

File: chapter_computational-performance/hybridize.md [hunk 3]
```diff
@@ -150,6 +184,11 @@ This seems almost too good to be true: simply designate a block to be `HybridSeq
...
+Explicitly adding the `jit_compile = True` flag to the `tf.function()` call enables XLA (Accelerated Linear Algebra) functionality in tensorflow. XLA can further optimize JIT compiled code in certain instances. Graph-mode execution is enabled without this explicit definition, however XLA can make certain large linear algebra operations (in the vein of those we see in deep learning applcations) much faster, particularly in a GPU environment.
```
Comment (terrytangyuan) [inline chapter_computational-performance/hybridize.md:189]:
```suggestion
Explicitly adding the `jit_compile = True` flag to the `tf.function()` call enables XLA (Accelerated Linear Algebra) functionality in TensorFlow. XLA can further optimize JIT compiled code in certain instances. Graph-mode execution is enabled without this explicit definition, however XLA can make certain large linear algebra operations (in the vein of those we see in deep learning applications) much faster, particularly in a GPU environment.
```

**Review by terrytangyuan — APPROVED** [review]
Thanks! This looks great and I just added a couple of small edits.

**Comment by mli** [comment]
Job d2l-en/PR-1646/5 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1646/

**Comment by levimcclenny** [comment]
CI is failing in pytorch scripts? Weird since it was passing before a couple grammatical errors corrected by @terrytangyuan. The failure doesn't appear to have anything to do with this PR. Any ideas here?

**Comment by mli** [comment]
Job d2l-en/PR-1646/7 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1646/

**Comment by levimcclenny** [comment]
Disregard, CI passed that time. Not sure what happened.

**Comment by terrytangyuan** [comment]
Thank you!

**Comment by astonzhang** [comment]
@levimcclenny  Thanks. Could you send a PR to replace your ID with name in the acknowledgement: https://github.com/d2l-ai/d2l-en/commit/d846f7d19ce526d07aa4fd7dfaf4d9bfbe1012f2
